### ISSUE TITLE
[Feature] Taylor expansion pruning

### ISSUE THREAD

**Issue by namgyu-youn** [pytorch/pytorch#157218]
### 🚀 The feature, motivation and pitch

Pruning (`torch.nn.utils.prune`) supports Ln-norm based pruning only using magnitude of weight.

More deeply, it works like the following:
```
Compute Ln-norm > Select low-weights using sort > Prune
```

And based on some research, Taylor expansion pruning can be helpful because it uses both weight and gradients. Full LaTeX can be checked at the "A Survey on Deep Neural Network Pruning—Section 5.4 (IEEE'24)". Following are the [arXiv](https://arxiv.org/abs/2308.06767) and section images:

![Image](https://github.com/user-attachments/assets/6273b982-6ae6-445c-a19d-03e1007ae50a)

How about introducing "first-order Taylor expansion pruning"? It seems that logic for computing gradient is only needed.

**Comment by namgyu-youn** [comment]
@jerryzh168 @jainapurva Could you check if this feature makes sense for `TorchAO`?

If it is, I want to open PR for this at `TorchAO`. Taylor expansion pruning might be placed in [torchao/prototype/sparsity/sparsifier/](https://github.com/pytorch/ao/tree/main/torchao/prototype/sparsity/sparsifier) using `BaseSparsifier`.
